### PR TITLE
Pipeline, Pavel Shkunov - 21FPL2

### PR DESCRIPTION
**Pull request by pa-shk**

### CODE REVIEW

File: lab_6_pipeline/pipeline.py [hunk 1]
```diff
@@ -18,22 +32,49 @@ def __init__(self, path_to_raw_txt_data: Path):
...
         """
         Validates folder with assets
         """
+        if not self.path_to_raw_txt_data.exists():
+            raise FileNotFoundError
+        if not self.path_to_raw_txt_data.is_dir():
+            raise NotADirectoryError
+        if not any(self.path_to_raw_txt_data.iterdir()):
+            raise  EmptyDirectoryError
+        meta_files = [i for i in self.path_to_raw_txt_data.glob('**/*.json') if  re.match(r'\d+_meta', i.stem)]
```
Comment (demid5111) [inline lab_6_pipeline/pipeline.py]: there should be no other json files or txt other than text ones

File: lab_6_pipeline/pipeline.py [hunk 2]
```diff
@@ -18,22 +32,49 @@ def __init__(self, path_to_raw_txt_data: Path):
...
 
     def _scan_dataset(self) -> None:
         """
         Register each dataset entry
         """
+        for path in self.path_to_raw_txt_data.glob('**/*.txt'):
+            if not (relevant:= re.search(r'(\d+)_raw', path.stem)):
+                continue
+            article = Article(url=None, article_id=int(relevant[1]))
+            from_raw(path, article)
```
Comment (demid5111) [inline lab_6_pipeline/pipeline.py]: from_raw creates article already

File: lab_6_pipeline/target_score.txt [hunk 3]
```diff
@@ -1 +1 @@
-0
+4
```
Comment (demid5111) [inline lab_6_pipeline/target_score.txt]: ?

File: lab_6_pipeline/pipeline.py [hunk 4]
```diff
@@ -144,17 +217,54 @@ def __init__(self, corpus_manager: CorpusManager):
...
         """
+        sentences = []
+        result = Mystem().analyze(re.sub(r'\W+', ' ', text))
+        mapping_path = Path(__file__).parent / 'data' / 'mystem_tags_mapping.json'
+        converter = MystemTagConverter(mapping_path)
+        token_count = 0
+        for sentence_position, sentence in enumerate(split_by_sentence(text)):
+            conllu_tokens = []
+            for token_position, token in enumerate(re.findall(r'\w+', sentence), start=1):
+                conlu_token = ConlluToken(token)
```
Comment (demid5111) [inline lab_6_pipeline/pipeline.py]: conllu_token

File: lab_6_pipeline/pipeline.py [hunk 5]
```diff
@@ -144,17 +217,54 @@ def __init__(self, corpus_manager: CorpusManager):
...
         Initializes MorphologicalAnalysisPipeline
         """
+        self._corpus = corpus_manager
 
     def _process(self, text: str) -> List[ConlluSentence]:
         """
         Returns the text representation as the list of ConlluSentence
         """
+        sentences = []
+        result = Mystem().analyze(re.sub(r'\W+', ' ', text))
```
Comment (demid5111) [inline lab_6_pipeline/pipeline.py]: mystem instance should be created once and stored as own attribute in init

File: lab_6_pipeline/pipeline.py [hunk 6]
```diff
@@ -144,17 +217,54 @@ def __init__(self, corpus_manager: CorpusManager):
...
+        mapping_path = Path(__file__).parent / 'data' / 'mystem_tags_mapping.json'
+        converter = MystemTagConverter(mapping_path)
```
Comment (demid5111) [inline lab_6_pipeline/pipeline.py]: the same with converter

**Review by demid5111 — CHANGES_REQUESTED** [review]
very nice, continue

File: lab_6_pipeline/pipeline.py (hunk 2 above)
Comment (pa-shk) [reply lab_6_pipeline/pipeline.py]: Fixed

File: lab_6_pipeline/pipeline.py (hunk 4 above)
Comment (pa-shk) [reply lab_6_pipeline/pipeline.py]: Fixed

File: lab_6_pipeline/pipeline.py (hunk 5 above)
Comment (pa-shk) [reply lab_6_pipeline/pipeline.py]: Fixed

File: lab_6_pipeline/pipeline.py (hunk 6 above)
Comment (pa-shk) [reply lab_6_pipeline/pipeline.py]: Fixed

File: lab_6_pipeline/pipeline.py (hunk 1 above)
Comment (pa-shk) [reply lab_6_pipeline/pipeline.py]: From lab desciption:

> NOTE: While validating dataset, ignore files which are named not using formats.

File: lab_6_pipeline/data/mystem_tags_mapping.json [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+        "им": "Nom",
+        "род": "Gen",
+        "дат": "Dat",
+        "вин": "Acc",
+        "твор": "Ins",
+        "пр": "Loc",
+        "зват": "Voc",
+        "парт": "Par",
+        "местн": "Loc",
+        "имя": "Acc"
```
Comment (demid5111) [inline lab_6_pipeline/data/mystem_tags_mapping.json]: where from do you get it?

File: lab_6_pipeline/data/opencorpora_tags_mapping.json [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+        "INTJ": "INTJ",
+        "UNKN": "X"
+    },
+    "Case": {
+        "nomn": "Nom",
+        "gent": "Gen",
+        "datv": "Dat",
+        "accs": "Acc",
+        "ablt": "Ins",
+        "loct": "Loc",
```
Comment (demid5111) [inline lab_6_pipeline/data/opencorpora_tags_mapping.json]: show any evidence that this is a correct mapping (official docs, our description, etc.)

File: lab_6_pipeline/pipeline.py [hunk 9]
```diff
@@ -18,22 +39,45 @@ def __init__(self, path_to_raw_txt_data: Path):
...
         """
         Validates folder with assets
         """
+        if not self.path_to_raw_txt_data.exists():
+            raise FileNotFoundError
+        if not self.path_to_raw_txt_data.is_dir():
+            raise NotADirectoryError
+        if not any(self.path_to_raw_txt_data.iterdir()):
+            raise  EmptyDirectoryError
+        meta_files = [i for i in self.path_to_raw_txt_data.glob('**/*.json') if  re.match(r'\d+_meta', i.stem)]
```
Comment (demid5111) [inline lab_6_pipeline/pipeline.py]: you put recursion in glob - why?

File: lab_6_pipeline/pipeline.py [hunk 10]
```diff
@@ -55,27 +101,50 @@ def __init__(self, text: str):
...
         """
         String representation of the token for conllu files
         """
+        position = str(self._position)
+        text = self._text
+        lemma = self._morphological_parameters.lemma
+        pos = self._morphological_parameters.pos
+        xpos = '_'
+        feats = '_'
+        if include_morphological_tags:
```
Comment (demid5111) [inline lab_6_pipeline/pipeline.py]: join conditions

File: lab_6_pipeline/pipeline.py [hunk 11]
```diff
@@ -86,22 +155,32 @@ def __init__(self, position: int, text: str, tokens: list[ConlluToken]):
...
         """
         Creates string representation of the sentence
         """
+        return self._format_tokens(include_morphological_tags)
 
     def get_cleaned_sentence(self) -> str:
         """
         Returns the lowercase representation of the sentence
         """
+        return ' '.join(filter(bool, (i.get_cleaned() for i in self._tokens)))
```
Comment (demid5111) [inline lab_6_pipeline/pipeline.py:178]: very nice filtering approach! 🏅

File: lab_6_pipeline/pipeline.py [hunk 12]
```diff
@@ -86,22 +155,32 @@ def __init__(self, position: int, text: str, tokens: list[ConlluToken]):
...
         """
+        self._position = position
+        self._text = text
+        self._tokens = tokens
 
     def get_conllu_text(self, include_morphological_tags: bool) -> str:
         """
         Creates string representation of the sentence
         """
+        return self._format_tokens(include_morphological_tags)
```
Comment (demid5111) [inline lab_6_pipeline/pipeline.py]: this method should create sentence block from conllu (id+sentence+words),self._format_tokens should just prepare words

File: lab_6_pipeline/pipeline.py [hunk 13]
```diff
@@ -112,11 +191,22 @@ def convert_morphological_tags(self, tags: str) -> str:  # type: ignore
         """
         Converts the Mystem tags into the UD format
         """
+        tags = re.sub(r'\((.+?)\|.+\)', r'\1', tags)
+        extracted_tags = re.findall(r'[а-я]+', tags)
+        ud_tags = {}
+        for tag in extracted_tags:
+            for category in (self.case, self.number, self.gender, self.animacy, self.tense):
```
Comment (demid5111) [inline lab_6_pipeline/pipeline.py]: conceptually this is overwork, for example for nouns you look for tense and so on. You'd better define some more strict rules on what features you extract for each pos

File: lab_6_pipeline/pipeline.py [hunk 14]
```diff
@@ -128,11 +218,18 @@ def convert_pos(self, tags: OpencorporaTagProtocol) -> str:  # type: ignore
...
         """
+        return self._tag_mapping[self.pos][tags.POS or 'UNKN']
 
     def convert_morphological_tags(self, tags: OpencorporaTagProtocol) -> str:  # type: ignore
         """
         Converts the OpenCorpora tags into the UD format
         """
+        ud_tags = {}
+        for category in ( 'animacy', 'case', 'gender', 'number'):
+            if not (open_corpora_tag := eval(f'tags.{category}')):
```
Comment (demid5111) [inline lab_6_pipeline/pipeline.py]: using eval is a bad practice, I would even ask to forget about it for 99.99% cases. rework

File: lab_6_pipeline/pipeline.py [hunk 15]
```diff
@@ -128,11 +218,18 @@ def convert_pos(self, tags: OpencorporaTagProtocol) -> str:  # type: ignore
...
+                continue
+            ud_tags[eval(f'self.{category}')] = eval(f"self._tag_mapping[self.{category}]['{open_corpora_tag}']")
```
Comment (demid5111) [inline lab_6_pipeline/pipeline.py]: come-on, lets say I have not seen so much evals in one place

File: lab_6_pipeline/pipeline.py [hunk 16]
```diff
@@ -144,17 +241,63 @@ def __init__(self, corpus_manager: CorpusManager):
         """
         Initializes MorphologicalAnalysisPipeline
         """
+        self._corpus = corpus_manager
+        self._stemmer = Mystem()
```
Comment (demid5111) [inline lab_6_pipeline/pipeline.py]: conceptually wrong name. we use it as analyzer and lemmatizer, not as a stemmer. btw, do you know the difference?

File: lab_6_pipeline/pipeline.py [hunk 17]
```diff
@@ -144,17 +241,63 @@ def __init__(self, corpus_manager: CorpusManager):
...
+        mapping_path = Path(__file__).parent / 'data' / 'mystem_tags_mapping.json'
+        self._converter = MystemTagConverter(mapping_path)
 
     def _process(self, text: str) -> List[ConlluSentence]:
         """
         Returns the text representation as the list of ConlluSentence
         """
+        sentences = []
+        result = self._stemmer.analyze(re.sub(r'\W+', ' ', text))
```
Comment (demid5111) [inline lab_6_pipeline/pipeline.py]: why do not you pass raw file content?

File: lab_6_pipeline/pipeline.py [hunk 18]
```diff
@@ -165,22 +308,77 @@ def __init__(self, corpus_manager: CorpusManager):
         """
         Initializes MorphologicalAnalysisPipeline
         """
+        super().__init__(corpus_manager)
+        self._backup_analyzer = pymorphy2.MorphAnalyzer()
```
Comment (demid5111) [inline lab_6_pipeline/pipeline.py:351]: and here you do not use names like stemmer, you see?

File: lab_6_pipeline/pipeline.py [hunk 19]
```diff
@@ -144,17 +241,63 @@ def __init__(self, corpus_manager: CorpusManager):
...
         """
+        sentences = []
+        result = self._stemmer.analyze(re.sub(r'\W+', ' ', text))
+        token_count = 0
+        for sentence_position, sentence in enumerate(split_by_sentence(text)):
+            conllu_tokens = []
+            for token_position, token in enumerate(re.findall(r'\w+', sentence), start=1):
+                conllu_token = ConlluToken(token)
+                if not result[token_count]['text'].isalnum():
+                    token_count += 1
```
Comment (demid5111) [inline lab_6_pipeline/pipeline.py]: why do you increase it here? and even allow it to be processed further. provide examples

File: lab_6_pipeline/pipeline.py [hunk 20]
```diff
@@ -165,22 +308,77 @@ def __init__(self, corpus_manager: CorpusManager):
...
+            sentences.append(sentence)
+        return sentences
 
     def run(self) -> None:
         """
         Performs basic preprocessing and writes processed text to files
         """
+        for article in self._corpus.get_articles().values():
+            article.set_conllu_sentences(self._process(article.text))
+            to_cleaned(article)
```
Comment (demid5111) [inline lab_6_pipeline/pipeline.py]: call is redundant

File: lab_6_pipeline/pipeline.py [hunk 21]
```diff
@@ -165,22 +308,77 @@ def __init__(self, corpus_manager: CorpusManager):
...
+        self._backup_analyzer = pymorphy2.MorphAnalyzer()
+        mapping_path = Path(__file__).parent / 'data' / 'opencorpora_tags_mapping.json'
+        self._backup_tag_converter = OpenCorporaTagConverter(mapping_path)
 
     def _process(self, text: str) -> List[ConlluSentence]:
         """
         Returns the text representation as the list of ConlluSentence
         """
+        sentences = []
+        result = self._stemmer.analyze(re.sub(r'\W+', ' ', text))
```
Comment (demid5111) [inline lab_6_pipeline/pipeline.py]: all comments from basic pipeline above apply here

File: lab_6_pipeline/pipeline.py [hunk 22]
```diff
@@ -165,22 +308,77 @@ def __init__(self, corpus_manager: CorpusManager):
...
+        token_count = 0
+        for sentence_position, sentence in enumerate(split_by_sentence(text)):
```
Comment (demid5111) [inline lab_6_pipeline/pipeline.py:361]: also think about large duplication between pipelines - how could you manage it? remember, API cannot be changed

File: lab_6_pipeline/pos_frequency_pipeline.py [hunk 23]
```diff
@@ -1,13 +1,40 @@
...
     Populates the Article abstraction with all information from the conllu file
     """
+    with open(path, encoding='utf-8') as f:
+        content = f.read()
+    extracted_sentences = extract_sentences_from_raw_conllu(content)
+    for sentence in extracted_sentences:
+        sentence['tokens'] = [_parse_conllu_token(token) for token in  sentence['tokens']]
+    conllu_senteces = [ConlluSentence(**sentence) for sentence in extracted_sentences]
+    if not article:
+        article_id = int(re.match(r'\d+', path.stem).group())
```
Comment (demid5111) [inline lab_6_pipeline/pos_frequency_pipeline.py]: use get_article_id_from_filepath to get id

File: lab_6_pipeline/pos_frequency_pipeline.py [hunk 24]
```diff
@@ -30,22 +67,45 @@ def __init__(self, corpus_manager: CorpusManager):
...
 
     def _count_frequencies(self, article: Article) -> dict[str, int]:
         """
         Counts POS frequency in Article
         """
+        frequencies = {}
+        for sentences in article.get_conllu_sentences():
+            for token in sentences.get_tokens():
+                pos = token.get_morphological_parameters().pos
+                frequencies[pos] = frequencies.get(pos, 0) + 1
```
Comment (demid5111) [inline lab_6_pipeline/pos_frequency_pipeline.py]: rework to use counter from collections

**Review by demid5111 — CHANGES_REQUESTED** [review]
very nice! especially good is POS pipeline. fix comments and you are almost done

File: lab_6_pipeline/data/mystem_tags_mapping.json (hunk 7 above)
Comment (pa-shk) [reply lab_6_pipeline/data/mystem_tags_mapping.json]: fixed

File: lab_6_pipeline/data/opencorpora_tags_mapping.json (hunk 8 above)
Comment (pa-shk) [reply lab_6_pipeline/data/opencorpora_tags_mapping.json]: fixed

File: lab_6_pipeline/pipeline.py (hunk 9 above)
Comment (pa-shk) [reply lab_6_pipeline/pipeline.py]: fixed

File: lab_6_pipeline/pipeline.py (hunk 10 above)
Comment (pa-shk) [reply lab_6_pipeline/pipeline.py]: fixed

File: lab_6_pipeline/pipeline.py (hunk 12 above)
Comment (pa-shk) [reply lab_6_pipeline/pipeline.py]: fixed

File: lab_6_pipeline/pipeline.py (hunk 13 above)
Comment (pa-shk) [reply lab_6_pipeline/pipeline.py]: fixed

File: lab_6_pipeline/pipeline.py (hunk 14 above)
Comment (pa-shk) [reply lab_6_pipeline/pipeline.py]: fixed

File: lab_6_pipeline/pipeline.py (hunk 15 above)
Comment (pa-shk) [reply lab_6_pipeline/pipeline.py]: fixed)

File: lab_6_pipeline/pipeline.py (hunk 16 above)
Comment (pa-shk) [reply lab_6_pipeline/pipeline.py]: Fixed. Yes, I do

File: lab_6_pipeline/pipeline.py (hunk 17 above)
Comment (pa-shk) [reply lab_6_pipeline/pipeline.py]: fixed

File: lab_6_pipeline/pipeline.py (hunk 18 above)
Comment (pa-shk) [reply lab_6_pipeline/pipeline.py:351]: yes, I see

File: lab_6_pipeline/pipeline.py (hunk 19 above)
Comment (pa-shk) [reply lab_6_pipeline/pipeline.py]: fixed

File: lab_6_pipeline/pipeline.py (hunk 20 above)
Comment (pa-shk) [reply lab_6_pipeline/pipeline.py]: fixed

File: lab_6_pipeline/pipeline.py (hunk 21 above)
Comment (pa-shk) [reply lab_6_pipeline/pipeline.py]: also fixed

File: lab_6_pipeline/pipeline.py (hunk 22 above)
Comment (pa-shk) [reply lab_6_pipeline/pipeline.py:361]: fixed

File: lab_6_pipeline/pos_frequency_pipeline.py (hunk 23 above)
Comment (pa-shk) [reply lab_6_pipeline/pos_frequency_pipeline.py]: fixed

File: lab_6_pipeline/pos_frequency_pipeline.py (hunk 24 above)
Comment (pa-shk) [reply lab_6_pipeline/pos_frequency_pipeline.py]: done

File: lab_6_pipeline/pipeline.py [hunk 25]
```diff
@@ -112,11 +198,34 @@ def convert_morphological_tags(self, tags: str) -> str:  # type: ignore
...
+        pos = self.convert_pos(tags)
+        if pos in ["PART", "ADP", "ADV", "CCONJ", "INTJ"]:
+            return ''
+
+        tags = re.sub(r'\((.+?)\|.+\)', r'\1', tags)
+        extracted_tags = re.findall(r'[а-я]+', tags)
+        ud_tags = {}
+        for tag in extracted_tags:
+            for category in (self.animacy, self.case, self.gender, self.number, self.tense):
+                if (pos == 'NOUN' or pos == 'PRON') and category == self.tense:
```
Comment (demid5111) [inline lab_6_pipeline/pipeline.py]: this is not a scalable approach. define rules of what attributes to extract for which pos.

File: lab_6_pipeline/pipeline.py [hunk 26]
```diff
@@ -144,16 +263,83 @@ def __init__(self, corpus_manager: CorpusManager):
...
+                else:
+                    tags, lex = '', token['text']
+                    pos = self._analyze_unknown(lex)
+                conllu_token = self._create_conllu_token(token_position, token['text'], lex, pos, tags)
+                conllu_tokens.append(conllu_token)
+            conllu_sentence = ConlluSentence(sentence_position, sentence, conllu_tokens)
+            conllu_sentences.append(conllu_sentence)
+        return conllu_sentences
+
+    @staticmethod
```
Comment (demid5111) [inline lab_6_pipeline/pipeline.py]: either make it bound, or extract to become a standalone method

File: lab_6_pipeline/pipeline.py [hunk 27]
```diff
@@ -144,16 +263,83 @@ def __init__(self, corpus_manager: CorpusManager):
...
+        """
+        Creates a ConlluToken and fills its fields
+        """
+        conllu_token = ConlluToken(text)
+        morph_params = MorphologicalTokenDTO(lex, pos, tags)
+        conllu_token.set_position(postion)
+        conllu_token.set_morphological_parameters(morph_params)
+        return conllu_token
+
+    @staticmethod
```
Comment (demid5111) [inline lab_6_pipeline/pipeline.py]: .

File: lab_6_pipeline/pipeline.py [hunk 28]
```diff
@@ -144,16 +263,83 @@ def __init__(self, corpus_manager: CorpusManager):
...
+        Deduces the part of spech of a lexeme
+        if analyzer failed to do it
+        """
+        if lex == '.':
+            return 'PUNCT'
+        if lex.isdigit():
+            return 'NUM'
+        return 'X'
+
+    @staticmethod
```
Comment (demid5111) [inline lab_6_pipeline/pipeline.py]: .

File: lab_6_pipeline/pos_frequency_pipeline.py [hunk 29]
```diff
@@ -17,6 +47,16 @@ def _parse_conllu_token(token_line: str) -> ConlluToken:
     Example:
     '2	произошло	происходить	VERB	_	Gender=Neut|Number=Sing|Tense=Past	0	root	_	_'
     """
+    params = token_line.split('\t')
+    position = int(params[0])
```
Comment (demid5111) [inline lab_6_pipeline/pos_frequency_pipeline.py]: use unpacking in several variables

**Review by demid5111 — CHANGES_REQUESTED** [review]
almost

File: lab_6_pipeline/pipeline.py (hunk 25 above)
Comment (pa-shk) [reply lab_6_pipeline/pipeline.py]: fixed

File: lab_6_pipeline/pipeline.py (hunk 26 above)
Comment (pa-shk) [reply lab_6_pipeline/pipeline.py]: fixed

File: lab_6_pipeline/pos_frequency_pipeline.py (hunk 29 above)
Comment (pa-shk) [reply lab_6_pipeline/pos_frequency_pipeline.py]: done

File: lab_6_pipeline/pipeline.py [hunk 30]
```diff
@@ -112,11 +196,35 @@ def convert_morphological_tags(self, tags: str) -> str:  # type: ignore
         """
         Converts the Mystem tags into the UD format
         """
+        pos = self.convert_pos(tags)
+
+        pos_categories = {
+            "NOUN": [self.case, self.number, self.gender, self.animacy],
```
Comment (demid5111) [inline lab_6_pipeline/pipeline.py:202]: nice! 🎖️

**Review by demid5111 — APPROVED** [review]
I am accepting your MR for the answer. make sure you have answered all my comments and can comment on them during oral handover.

**Pull request closed without merging**